pkg/controller/controller_instance.go

```diff
@@ -303,15 +337,43 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) error {
 	glog.V(4).Infof("Provisioning a new Instance %v/%v of ServiceClass %v at Broker %v", instance.Namespace, instance.Name, serviceClass.Name, brokerName)
 	response, err := brokerClient.ProvisionInstance(request)
 	if err != nil {
+		// There are two buckets of errors to handle:
+		// 1.  Errors that represent a failure response from the broker
+		// 2.  All other errors
+		if osb.IsHTTPError(err) {
+			// An error from the broker represents a permanent failure and
+			// should not be retried; set the Failed condition.
+
+			httpErr := err.(osb.HTTPStatusCodeError)
+			s := fmt.Sprintf("Error provisioning Instance \"%s/%s\" of ServiceClass %q at Broker %q: %s", instance.Namespace, instance.Name, serviceClass.Name, brokerName, httpErr)
+			glog.Warning(s)
+			clone, err := api.Scheme.DeepCopy(instance)
+			if err != nil {
+				return err
+			}
+			toUpdate := clone.(*v1alpha1.Instance)
+
+			setInstanceCondition(toUpdate, v1alpha1.InstanceConditionFailed, v1alpha1.ConditionTrue, "BrokerReturnedFailure", s)
+			c.updateInstanceCondition(
```

Comment:
Seeing set and then update confuses me. Update calls set inside, but even so, it seems weird to me.

we set failed on the clone, and not ready on the original. I'm a little lost reading through this. Might need some more comments. I need to think on this some more and read it again.